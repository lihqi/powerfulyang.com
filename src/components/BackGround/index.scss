canvas {
  position:fixed;
  top:0;
  left:0;
  z-index: 1;
}
.bg{
  @extend canvas;
  width: 100%;
  height: 100%;
  background-size:100% 100%;
  background-image: url(/IMG_4405.png);
}
