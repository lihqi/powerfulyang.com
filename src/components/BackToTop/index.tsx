import RocketShip from '@/lottie/RocketShip.json';
import Lottie from 'lottie-react';
import React from 'react';
import { useScrollDirection } from '@powerfulyang/hooks';
import { AnimatePresence, motion } from 'framer-motion';

export const BackToTop = () => {
  const { isDown } = useScrollDirection();
  return (
    <AnimatePresence>
      {isDown && (
        <motion.div
          title="Back to top"
          className="pointer fixed bottom-28 right-10 w-28"
          transition={{
            duration: 0.3,
            type: 'keyframes',
            ease: 'easeOut',
          }}
          initial={{
            opacity: 0,
            y: '-80px',
          }}
          animate={{
            opacity: 1,
            y: 0,
          }}
          exit={{
            opacity: 0,
            y: '-80vh',
          }}
        >
          <Lottie
            onClick={() => {
              window.scrollTo({
                top: 0,
                behavior: 'smooth',
              });
            }}
            animationData={RocketShip}
          />
        </motion.div>
      )}
    </AnimatePresence>
  );
};
