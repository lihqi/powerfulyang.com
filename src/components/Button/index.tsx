import React, { FC } from 'react';
import classNames from 'classnames';
import { ButtonProps } from '@/components/Button/types/BaseButtonProps';
import { Loading } from '@/components/Loading';
import styles from './index.module.scss';

export const Button: FC<ButtonProps> = (props) => {
  const { htmlType = 'button', children, className, icon, loading = false, ...rest } = props;
  return (
    <button {...rest} type={htmlType} className={classNames(className, styles.button)}>
      {icon}
      {loading && <Loading />}
      {children}
    </button>
  );
};

Button.displayName = 'Button';
