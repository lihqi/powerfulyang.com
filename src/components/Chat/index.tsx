import type { ChatMessageEntity, Message } from '@/components/Chat/Message';
import { ChatMessage } from '@/components/Chat/Message';
import type { FC } from 'react';
import { useState } from 'react';
import { handlePasteImageAndReturnFileList } from '@/utils/copy';
import styles from './index.module.scss';

export type SentMessage = Message & { fileType?: string };

type Props = {
  messages: ChatMessageEntity[];
  onSendMessage: (message: SentMessage) => void;
};

export const Chat: FC<Props> = ({ messages, onSendMessage }) => {
  const [message, setMessage] = useState('');
  return (
    <div className={styles.chat}>
      <div className={styles.messages}>
        {messages.map((v) => (
          <ChatMessage key={v.messageId} {...v} />
        ))}
      </div>
      <textarea
        onChange={(e) => {
          setMessage(e.target.value);
        }}
        value={message}
        placeholder="请输入消息..."
        className={styles.chatInput}
        onKeyDown={(e) => {
          if (e.key === 'Enter' && (e.ctrlKey || e.metaKey)) {
            setMessage(`${message}\n`);
          }
          if (e.key === 'Enter' && !(e.ctrlKey || e.metaKey)) {
            e.preventDefault();
            onSendMessage({ messageContentType: 'text', content: message });
            setMessage('');
          }
        }}
        onPaste={(e) => {
          const fileList = handlePasteImageAndReturnFileList(e);
          if (fileList) {
            for (const file of fileList) {
              const reader = new FileReader();
              reader.onload = () => {
                const blob = new Blob([reader.result as ArrayBuffer], {
                  type: file.type,
                });
                onSendMessage({
                  messageContentType: 'blob',
                  content: blob,
                  fileType: file.type,
                });
              };
              reader.readAsArrayBuffer(file);
            }
          }
        }}
      />
    </div>
  );
};
