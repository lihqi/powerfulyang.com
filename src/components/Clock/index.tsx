import React, { FC, HTMLProps, useEffect, useState } from 'react';
import { interval } from 'rxjs';
import classNames from 'classnames';
import { TimeFormat } from '@/utils/lib';
import styles from './index.module.scss';

export const Clock: FC<HTMLProps<HTMLDivElement>> = ({ className }) => {
  const [time, setTime] = useState<string>();
  const [start, setStart] = useState(false);
  useEffect(() => {
    const sub = interval().subscribe(() => {
      setTime(TimeFormat());
    });
    return () => {
      sub.unsubscribe();
    };
  }, []);
  useEffect(() => {
    if (time?.[5] === '0') {
      setStart(true);
    }
  }, [time]);
  return (
    <div className={classNames(styles.clock, className)}>
      {time && (
        <>
          <img src={`/numbers/${time[0]}.gif`} alt="" />
          <img src={`/numbers/${time[1]}.gif`} alt="" />
          <img src={`/numbers/${time[2]}.gif`} alt="" />
          <img src={`/numbers/${time[3]}.gif`} alt="" />
          <img
            src={`/numbers/${time[4]}.gif`}
            alt=""
            className={classNames({
              [styles.rotate]: start,
            })}
          />
          <img src={`/numbers/${time[5]}.gif`} alt="" />
        </>
      )}
    </div>
  );
};
