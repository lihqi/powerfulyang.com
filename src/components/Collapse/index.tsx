import React, { FC, useEffect, useRef, useState } from 'react';
import classNames from 'classnames';
import styles from './index.module.scss';

type CollapseProps = {
  collapsed: boolean;
};

export const Collapse: FC<CollapseProps> = ({ children, collapsed }) => {
  const ref = useRef(0);
  const divRef = useRef<HTMLDivElement>(null);
  const [renderClass, setRenderClass] = useState('');
  useEffect(() => {
    ref.current++;
  }, []);
  useEffect(() => {
    const div = divRef.current!;
    if (collapsed) {
      if (ref.current === 0) {
        // first render
        setRenderClass('h-auto overflow-visible visible');
      } else {
        div.style.height = `${div.scrollHeight}px`;
      }
    } else {
      div.style.height = `0px`;
      setRenderClass('overflow-hidden');
    }
  }, [collapsed]);
  return (
    <div ref={divRef} className={classNames(renderClass, styles.collapse)}>
      {children}
    </div>
  );
};
