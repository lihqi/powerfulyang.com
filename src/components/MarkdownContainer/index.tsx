'use client';

import type { MarkdownMetadata } from '@/components/MarkdownContainer/LiveMarkdownEditor';
import { remarkDirectiveHandle } from '@/components/MarkdownContainer/plugins/remarkDirectiveHandle';
import { remarkMetadata } from '@/components/MarkdownContainer/plugins/remarkMetadata';
import { copyToClipboardAndNotify } from '@/utils/copy';
import { Icon } from '@powerfulyang/components';
import classNames from 'classnames';
// `rehype-katex` does not import the CSS for you
import 'katex/dist/katex.min.css';
import type { FC } from 'react';
import React from 'react';
import ReactMarkdown from 'react-markdown';
import rehypeKatex from 'rehype-katex';
import rehypeSlug from 'rehype-slug';
import remarkDirective from 'remark-directive';
import remarkFrontmatter from 'remark-frontmatter';
import remarkGfm from 'remark-gfm';
import remarkMath from 'remark-math';

import styles from './index.module.scss';
import { A, BlockQuote, Code, H1, H2, H3, H4, Img, Li, Pre, Table, Ul } from './MarkdownElement';
import rehypeRaw from './plugins/rehypeRaw';

export type MarkdownContainerProps = {
  source: string;
  className?: string;
  metadataRef?: React.MutableRefObject<MarkdownMetadata>;
};

export const MarkdownContainer: FC<MarkdownContainerProps> = ({
  source,
  className,
  metadataRef,
}) => {
  return (
    <ReactMarkdown
      className={classNames(styles.markdownBody, className)}
      remarkPlugins={[
        remarkGfm,
        remarkFrontmatter,
        [remarkMetadata, metadataRef],
        remarkMath,
        remarkDirective,
        remarkDirectiveHandle,
      ]}
      remarkRehypeOptions={{
        // @ts-ignore
        passThrough: ['info', 'tags'],
        handlers: {
          // @ts-ignore
          info(state, node) {
            return {
              type: 'element',
              tagName: 'div',
              properties: {
                className: styles.postInfo,
              },
              children: state.all(node),
            };
          },
          // @ts-ignore
          tags(state, node) {
            return {
              type: 'element',
              tagName: 'tags',
              properties: {},
              children: state.all(node),
            };
          },
        },
      }}
      rehypePlugins={[rehypeSlug, rehypeKatex, rehypeRaw]}
      components={{
        h1: H1,
        h2: H2,
        h3: H3,
        h4: H4,
        blockquote: BlockQuote,
        table: Table,
        a: A,
        code: Code,
        pre: Pre,
        li: Li,
        ul: Ul,
        img: Img,
        // @ts-ignore
        // eslint-disable-next-line react/no-unstable-nested-components
        tags({ children }) {
          const tags = Array.isArray(children) ? children : [children];
          return (
            <div className="flex flex-wrap sm:ml-2 lg:ml-6">
              {tags?.map((tag: string) => (
                <button
                  type="button"
                  key={tag}
                  className="pointer my-2 mr-2"
                  onClick={() => copyToClipboardAndNotify(tag)}
                >
                  <Icon type="icon-tag" className="text-xl" />
                  <span className="text-sm text-[#FFB356]">{tag}</span>
                </button>
              ))}
            </div>
          );
        },
      }}
    >
      {source}
    </ReactMarkdown>
  );
};

export default MarkdownContainer;
