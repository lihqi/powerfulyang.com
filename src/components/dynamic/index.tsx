import dynamic from 'next/dynamic';

/**
 * @deprecated
 */
export const TwitterFavoriteWithNoSSR = dynamic(
  () => import('@/components/dynamic/TwitterFavorite'),
  {
    ssr: false,
  },
);

/**
 * @deprecated
 */
export const MapleLeafFallingWithNoSSR = dynamic(
  () => import('@/deprecated/unSupportSSR/MapleLeafFalling'),
  { ssr: false },
);

export const TagCloudWithNoSSR = dynamic(() => import('./TagCloud'), { ssr: false });
