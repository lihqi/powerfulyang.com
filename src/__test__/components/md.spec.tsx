import { MarkdownContainer } from '@/components/MarkdownContainer';
import { describe, expect, it } from '@jest/globals';
import { render } from '@testing-library/react';

describe('md', () => {
  it('render', () => {
    const md = `# 2022\n## 2022-01-01\n发布了 [2022-01-01](/post/2022-01-01)。`;
    const result = render(<MarkdownContainer source={md} />);
    expect(result.container).toMatchSnapshot();
  });
});
