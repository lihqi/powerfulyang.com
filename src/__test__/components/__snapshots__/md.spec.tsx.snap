// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`md render 1`] = `
<div>
  <div
    class="relative"
  >
    <div
      id="2022"
      title="anchor"
    />
    <h1
      class="flex w-full justify-center pb-2"
    >
      <span>
        <span />
        <span>
          2022
        </span>
        <span />
      </span>
    </h1>
  </div>
  

  <div
    class="relative"
  >
    <div
      id="2022-01-01"
      title="anchor"
    />
    <h2
      class="cursor-text"
    >
      2022-01-01
    </h2>
  </div>
  

  <p>
    发布了 
    <a
      class="link"
      href="/post/2022-01-01"
      rel="noreferrer"
      target="_blank"
    >
      2022-01-01
    </a>
    。
  </p>
</div>
`;
